src/main/scala/firrtl/passes/NewLowerTypes.scala

```diff
@@ -0,0 +1,454 @@
+// See LICENSE for license details.
+
+package firrtl.passes
+
+import firrtl.analyses.InstanceKeyGraph
+import firrtl.annotations.{CircuitTarget, MemoryInitAnnotation, MemoryRandomInitAnnotation, ModuleTarget, ReferenceTarget}
+import firrtl.{CircuitForm, CircuitState, InstanceKind, Kind, MemKind, PortKind, RenameMap, SymbolTable, Transform, UnknownForm, Utils}
+import firrtl.ir._
+import firrtl.options.Dependency
+import firrtl.stage.TransformManager.TransformDependency
+
+import scala.collection.mutable
+
+/** Flattens Bundles and Vecs.
+  * - Some implicit bundle types remain, but with a limited depth:
+  *   - the type of a memory is still a bundle with depth 2 (mem -> port -> field), see [[MemPortUtils.memType]]
+  *   - the type of a module instance is still a bundle with depth 1 (instance -> port)
+  */
+object NewLowerTypes extends Transform {
+  override def inputForm: CircuitForm = UnknownForm
+  override def outputForm: CircuitForm = UnknownForm
```

Comment:
Mixin `with DependencyAPIMigration` and remove deprecated `inputForm`/`outputForm`.

Reply:
Will do. I didn't do this initially since the word *Migration* suggested to me that this is only for migrating legacy passes.